snoseeds/Banka
# #165257507 API endpoint for user to create a bank account

#### What does this PR do?
Implements API endpoint for client sign up at `host/api/v1/accounts`
API endpoint for user sign in
#### Description of Task to be completed?
- Refactoring `validateFormFields` function to take an object that has key-value pairs of the description of required fields and their form / variable names.
- Using TDD to write detailed expected behavior of the `api` for a user that just signed up or that has signed in, which first fails and then passes eventually by doing the below actions.
- Getting the token from the authorization property of the header object of the request object.
- Passing the token to a function `authenticateUser` that verifies if the `type of user` from the token's payload is `client`, (else, a 403 response is returned with an error message) then checking if the `email` from the token's payload exists in the database (else, it returns a 400 response with an error message).
- Returning 400 status code and JSON object with relevant error messages when any of `accountType` (savings or current), `idCardType`, or `idCardNumber` is missing in the request, else,
- Passing unique information of the user to a function: `generateAccountNo`, which uses a novel algorithm to generate a unique account number for the user.
- Constructing a new account for the user with relevant information including the newly generated account number, and updating the database as appropriate
- Returning a 201 status code with JSON object having a JSend format with relevant information of the user and the newly created account.
#### How should this be manually tested?
- Running `npm test` on the command line, or by sending post requests through Postman to actually see the response status and error or data objects under various queries to the server as shown in the TDD tests written.
- With this being a protected route, the post request needs authorization property from Postman, and the authorization property is to be gotten by first sending a sign up or sign in request, copying the token from the response received, and pasting that token into the token field of the `Bearer Token` option under authorization of the post request to test the `api` in question.
#### What are the relevant pivotal tracker stories?
[API endpoint for user to create a bank account](https://www.pivotaltracker.com/story/show/165257507)
